File: app/src/main/java/com/ebnbin/gank/feature/days/day/DayFragment.java
package com.ebnbin.gank.feature.days.day;

import android.os.Bundle;
import android.support.annotation.CallSuper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.ebnbin.eb.util.EBUtil;
import com.ebnbin.eb.util.Timestamp;
import com.ebnbin.ebapplication.context.EBActionBarFragment;
import com.ebnbin.ebapplication.context.EBFragment;
import com.ebnbin.ebapplication.net.NetModelCallback;
import com.ebnbin.gank.R;

import java.util.ArrayList;
import java.util.Locale;

import okhttp3.Call;
import okhttp3.Response;

/**
 * 用 {@link RecyclerView} 展示某日期的数据.
 */
public final class DayFragment extends EBFragment {
    //*****************************************************************************************************************
    // Arguments.

    private static final String ARG_TIMESTAMP = "timestamp";

    @NonNull
    public static DayFragment newInstance(@NonNull Timestamp timestamp) {
        Bundle args = new Bundle();
        args.putParcelable(ARG_TIMESTAMP, timestamp);

        DayFragment dayFragment = new DayFragment();
        dayFragment.setArguments(args);

        return dayFragment;
    }

    private Timestamp mTimestamp;

    @Override
    protected void onInitArguments(@NonNull Bundle args) {
        super.onInitArguments(args);

        mTimestamp = args.getParcelable(ARG_TIMESTAMP);
    }

    //*****************************************************************************************************************

    private RecyclerView mDayRecyclerView;

    @Override
    protected int overrideContentViewLayout() {
        return R.layout.days_day_fragment;
    }

    private DayLayoutManager mLayoutManager;
    private DayAdapter mAdapter;
    private DayItemDecoration mItemDecoration;

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mDayRecyclerView = view.findViewById(R.id.day);

        mLayoutManager = new DayLayoutManager(getContext());
        mDayRecyclerView.setLayoutManager(mLayoutManager);

        mAdapter = new DayAdapter();
        mAdapter.listeners.add(new DayAdapter.Listener() {
            @Override
            void onDataClick(@NonNull DayEntity.Data data) {
                super.onDataClick(data);

                if (data.dataModel.getValidUrl() != null) {
                    getEbActivity().loadUrl(data.dataModel.getValidUrl());
                } else {
                    // TODO
                }
            }
        });
        mDayRecyclerView.setAdapter(mAdapter);

        mItemDecoration = new DayItemDecoration(getContext());
        mDayRecyclerView.addItemDecoration(mItemDecoration);

        mDayRecyclerView.setItemViewCacheSize(32);
    }

    public RecyclerView getDayRecyclerView() {
        return mDayRecyclerView;
    }

    //*****************************************************************************************************************
    // Instance state.

    private static final String STATE_DAY_MODEL = "day_model";

    @Override
    public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
        super.onViewStateRestored(savedInstanceState);

        if (mTimestamp == null) {
            // TODO: Sets empty view.
            return;
        }

        if (!restoreDay(savedInstanceState)) {
            netGetDay();
        }
    }

    /**
     * 恢复 {@link #mDayModel}.
     *
     * @return 是否成功恢复.
     */
    private boolean restoreDay(@Nullable Bundle savedInstanceState) {
        if (savedInstanceState == null) {
            return false;
        }

        mDayModel = (DayModel) savedInstanceState.getSerializable(STATE_DAY_MODEL);
        if (mDayModel == null) {
            return false;
        }

        mAdapter.setDay(mDayModel);

        return true;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        if (outState == null) {
            return;
        }

        outState.putSerializable(STATE_DAY_MODEL, mDayModel);
    }

    //*****************************************************************************************************************
    // Net.

    private DayModel mDayModel;

    public void getDayModel(@NonNull DayModelCallback callback) {
        if (mDayModel == null) {
            mDayModelCallbackArrayList.add(callback);
        } else {
            callback.onGetDayModel(mDayModel);
        }
    }

    private final ArrayList<DayModelCallback> mDayModelCallbackArrayList = new ArrayList<>();

    public static abstract class DayModelCallback {
        public void onGetDayModel(@NonNull DayModel dayModel) {
        }
    }

    /**
     * Gets {@link DayModel} model and sets data.
     */
    private void netGetDay() {
        String url = getDayUrl();
        netGet(url, new NetModelCallback<DayModel>() {
            @Override
            public void onSuccess(@NonNull Call call, @NonNull DayModel model, @NonNull Response response,
                    @NonNull byte[] byteArray) {
                super.onSuccess(call, model, response, byteArray);

                if (getUserVisibleHint()) {
                    EBActionBarFragment actionBarFragment = getActionBarParentFragment();
                    if (actionBarFragment != null) {
                        actionBarFragment.setActionBarMode(EBActionBarFragment.ActionBarMode.SCROLL, true, true, true);
                    }
                }

                mDayModel = model;

                // TODO java.util.ConcurrentModificationException
                for (DayModelCallback dayModelCallback : mDayModelCallbackArrayList) {
                    dayModelCallback.onGetDayModel(mDayModel);

                    mDayModelCallbackArrayList.remove(dayModelCallback);
                }

                mAdapter.setDay(mDayModel);
//
//                // Preload 福利 image.
//                if (!TextUtils.isEmpty(mDayModel.getResults().getValidFuliUrl())) {
//                    Picasso
//                            .with(getContext())
//                            .load(mDayModel.getResults().getValidFuliUrl())
//                            .fetch();
//                }
            }
        });
    }

    @CallSuper
    @Override
    protected void onFront() {
        super.onFront();

        EBUtil.INSTANCE.getHandler().post(new Runnable() {
            @Override
            public void run() {
                if (mDayRecyclerView == null) {
                    EBUtil.INSTANCE.getHandler().postDelayed(this, 16L);

                    return;
                }

                EBActionBarFragment actionBarFragment = getActionBarParentFragment();
                if (actionBarFragment != null) {
                    actionBarFragment.setNestedScrollingChild(mDayRecyclerView);
                    actionBarFragment.setActionBarMode(EBActionBarFragment.ActionBarMode.SCROLL, false, null, false);
                }
            }
        });
    }

    /**
     * Gets url of getting {@link DayModel}.
     */
    private String getDayUrl() {
        assert mTimestamp != null;

        String dayFormat = "http://gank.io/api/day/%04d/%02d/%02d";
        return String.format(Locale.getDefault(), dayFormat, mTimestamp.getYear(), mTimestamp.getMonth(),
                mTimestamp.getDay());
    }
}

File: app/src/main/java/com/ebnbin/gank/feature/days/day/DayFragment.kt
package com.ebnbin.gank.feature.days.day

import android.os.Bundle
import android.support.annotation.CallSuper
import android.support.v7.widget.RecyclerView
import android.view.View
import com.ebnbin.eb.util.EBUtil
import com.ebnbin.eb.util.Timestamp
import com.ebnbin.ebapplication.context.EBActionBarFragment
import com.ebnbin.ebapplication.context.EBFragment
import com.ebnbin.ebapplication.net.NetModelCallback
import com.ebnbin.gank.R
import com.ebnbin.gank.feature.data.DataAdapter
import com.ebnbin.gank.feature.data.DataEntity
import com.ebnbin.gank.feature.data.DataItemDecoration
import com.ebnbin.gank.feature.data.DataLayoutManager
import okhttp3.Call
import okhttp3.Response
import java.util.*

/**
 * 用 [RecyclerView] 展示某日期的数据.
 */
class DayFragment : EBFragment() {
    private val timestamp: Timestamp by lazy {
        arguments.getParcelable<Timestamp>(ARG_TIMESTAMP)
    }

    //*****************************************************************************************************************

    override fun overrideContentViewLayout(): Int {
        return R.layout.days_day_fragment
    }

    private val dayRecyclerView: RecyclerView by lazy {
        stateView.findViewById(R.id.day) as RecyclerView
    }

    private val layoutManager: DataLayoutManager by lazy {
        DataLayoutManager(context)
    }

    private val adapter: DataAdapter by lazy {
        DataAdapter()
    }

    private val itemDecoration: DataItemDecoration by lazy {
        DataItemDecoration(context)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        dayRecyclerView.layoutManager = layoutManager
        adapter.listeners.add(object : DataAdapter.Listener() {
            override fun onDataClick(data: DataEntity.Data) {
                super.onDataClick(data)

                if (data.dataModel.validUrl != null) {
                    ebActivity.loadUrl(data.dataModel.validUrl!!)
                }
            }
        })
        dayRecyclerView.adapter = adapter
        dayRecyclerView.addItemDecoration(itemDecoration)
        dayRecyclerView.setItemViewCacheSize(32)
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        super.onViewStateRestored(savedInstanceState)

        if (!restoreDay(savedInstanceState)) {
            netGetDay()
        }
    }

    /**
     * 恢复 [dayModel].

     * @return 是否成功恢复.
     */
    private fun restoreDay(savedInstanceState: Bundle?): Boolean {
        if (savedInstanceState == null) {
            return false
        }

        dayModel = savedInstanceState.getSerializable(STATE_DAY_MODEL) as DayModel
        if (dayModel == null) {
            return false
        }

        adapter.setNewData(dayModel)

        return true
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)

        if (outState == null) {
            return
        }

        outState.putSerializable(STATE_DAY_MODEL, dayModel)
    }

    //*****************************************************************************************************************
    // Net.

    private var dayModel: DayModel? = null

    fun getDayModel(callback: DayModelCallback) {
        if (dayModel == null) {
            dayModelCallbacks.add(callback)
        } else {
            callback.onGetDayModel(dayModel!!)
        }
    }

    private val dayModelCallbacks = ArrayList<DayModelCallback>()

    abstract class DayModelCallback {
        open fun onGetDayModel(dayModel: DayModel) {}
    }

    private val dayUrl: String by lazy {
        "http://gank.io/api/day/%04d/%02d/%02d".format(timestamp.year, timestamp.month, timestamp.day)
    }

    /**
     * Gets [DayModel] model and sets data.
     */
    private fun netGetDay() {
        val url = dayUrl
        netGet(url, object : NetModelCallback<DayModel>() {
            override fun onSuccess(call: Call, model: DayModel, response: Response,
                    byteArray: ByteArray) {
                super.onSuccess(call, model, response, byteArray)

                if (userVisibleHint) {
                    actionBarParentFragment?.setActionBarMode(EBActionBarFragment.ActionBarMode.SCROLL, true, true,
                            true)
                }

                dayModel = model

                for (dayModelCallback in dayModelCallbacks) {
                    dayModelCallback.onGetDayModel(dayModel!!)
                }
                dayModelCallbacks.clear()

                adapter.setNewData(dayModel)

                // TODO Preload 福利 image.
            }
        })
    }

    @CallSuper override fun onFront() {
        super.onFront()

        EBUtil.handler.post {
            if (actionBarParentFragment != null) {
                actionBarParentFragment!!.setNestedScrollingChild(dayRecyclerView)
                actionBarParentFragment!!.setActionBarMode(EBActionBarFragment.ActionBarMode.SCROLL, false, null,
                        false)
            }
        }
    }

    companion object {
        //*************************************************************************************************************
        // Arguments.

        private val ARG_TIMESTAMP = "timestamp"

        fun newInstance(timestamp: Timestamp): DayFragment {
            val args = Bundle()
            args.putParcelable(ARG_TIMESTAMP, timestamp)

            val dayFragment = DayFragment()
            dayFragment.arguments = args

            return dayFragment
        }

        //*************************************************************************************************************
        // Instance state.

        private val STATE_DAY_MODEL = "day_model"
    }
}
